community/io/src/main/java/org/neo4j/io/mem/GrabAllocator.java

```diff
@@ -82,8 +82,21 @@ public synchronized long allocateAligned( long bytes, long alignment )
     {
         if ( alignment <= 0 )
         {
-            throw new IllegalArgumentException( "Invalid alignment: " + alignment + "; alignment must be positive" );
+            throw new IllegalArgumentException( "Invalid alignment: " + alignment + ". Alignment must be positive." );
         }
+        try
+        {
+            return innerAllocateAligned( bytes, alignment );
+        }
+        catch ( NativeMemoryAllocationRefusedError nmare )
+        {
+            nmare.setAlreadyAllocatedBytes( usedMemory() );
```

Comment:
not quite sure why do we need to post initialize exception in this way, track its state internally inside of it and change message based on it instead of doing something like ():
```
Index: community/io/src/main/java/org/neo4j/io/mem/GrabAllocator.java
@@ -84,15 +84,7 @@
         {
             throw new IllegalArgumentException( "Invalid alignment: " + alignment + ". Alignment must be positive." );
         }
-        try
-        {
-            return innerAllocateAligned( bytes, alignment );
-        }
-        catch ( NativeMemoryAllocationRefusedError nmare )
-        {
-            nmare.setAlreadyAllocatedBytes( usedMemory() );
-            throw nmare;
-        }
+        return innerAllocateAligned( bytes, alignment );
     }
 
     private long innerAllocateAligned( long bytes, long alignment )
@@ -101,11 +93,11 @@
         {
             // This is a huge allocation. Put it in its own grab and keep any existing grab at the head.
             Grab nextGrab = grabs == null ? null : grabs.next;
-            Grab allocationGrab = new Grab( nextGrab, bytes );
+            Grab allocationGrab = new Grab( this, nextGrab, bytes );
             if ( !allocationGrab.canAllocate( bytes ) )
             {
                 allocationGrab.free();
-                allocationGrab = new Grab( nextGrab, bytes + alignment );
+                allocationGrab = new Grab( this, nextGrab, bytes + alignment );
             }
             long allocation = allocationGrab.allocate( bytes, alignment );
             grabs = grabs == null ? allocationGrab : grabs.setNext( allocationGrab );
@@ -119,7 +111,7 @@
             if ( desiredGrabSize < bytes )
             {
                 desiredGrabSize = bytes;
-                Grab grab = new Grab( grabs, desiredGrabSize );
+                Grab grab = new Grab( this, grabs, desiredGrabSize );
                 if ( grab.canAllocate( bytes ) )
                 {
                     memoryReserve -= desiredGrabSize;
@@ -130,7 +122,7 @@
                 desiredGrabSize = bytes + alignment;
             }
             memoryReserve -= desiredGrabSize;
-            grabs = new Grab( grabs, desiredGrabSize );
+            grabs = new Grab( this, grabs, desiredGrabSize );
         }
         return grabs.allocate( bytes, alignment );
     }
@@ -148,7 +140,7 @@
         }
     }
 
-    private static long allocateNativeMemory( long size )
+    private static long allocateNativeMemory( GrabAllocator grabAllocator, long size )
     {
         try
         {
@@ -156,7 +148,7 @@
         }
         catch ( OutOfMemoryError e )
         {
-            NativeMemoryAllocationRefusedError error = new NativeMemoryAllocationRefusedError( size );
+            NativeMemoryAllocationRefusedError error = new NativeMemoryAllocationRefusedError( size, grabAllocator.usedMemory() );
             try
             {
                 error.initCause( e );
@@ -187,10 +179,10 @@
         private final long limit;
         private long nextPointer;
 
-        Grab( Grab next, long size )
+        Grab( GrabAllocator allocator, Grab next, long size )
         {
             this.next = next;
-            this.address = allocateNativeMemory( size );
+            this.address = allocateNativeMemory( allocator, size);
             this.limit = address + size;
 
             nextPointer = address;
Index: community/io/src/main/java/org/neo4j/io/mem/NativeMemoryAllocationRefusedError.java
@@ -27,12 +27,12 @@
 public class NativeMemoryAllocationRefusedError extends OutOfMemoryError
 {
     private final long attemptedAllocationSizeBytes;
-    private boolean hasAlreadyAllocatedBytes;
-    private long alreadyAllocatedBytes;
+    private final long alreadyAllocatedBytes;
 
-    public NativeMemoryAllocationRefusedError( long size )
+    NativeMemoryAllocationRefusedError( long size, long alreadyAllocatedBytes )
     {
         this.attemptedAllocationSizeBytes = size;
+        this.alreadyAllocatedBytes = alreadyAllocatedBytes;
     }
 
     @Override
@@ -41,11 +41,8 @@
         String message = super.getMessage();
         StringBuilder sb = new StringBuilder();
         sb.append( "Failed to allocate " ).append( attemptedAllocationSizeBytes ).append( " bytes. " );
-        if ( hasAlreadyAllocatedBytes )
-        {
-            sb.append( "So far " ).append( alreadyAllocatedBytes );
-            sb.append( " bytes have already been successfully allocated. " );
-        }
+        sb.append( "So far " ).append( alreadyAllocatedBytes );
+        sb.append( " bytes have already been successfully allocated. " );
 
         sb.append( "The system currently has " );
         appendBytes( sb, getTotalPhysicalMemory() ).append( " total physical memory, " );
@@ -76,10 +73,4 @@
         }
         return sb;
     }
-
-    public void setAlreadyAllocatedBytes( long alreadyAllocatedBytes )
-    {
-        hasAlreadyAllocatedBytes = true;
-        this.alreadyAllocatedBytes = alreadyAllocatedBytes;
-    }
 }
```